File: src/main/python/AoC2021_16.py
#! /usr/bin/env python3
#
# Advent of Code 2021 Day 16
#

from math import prod
from aoc import my_aocd
from aoc.common import log
import aocd


def _hex2bin(hex_data: str) -> str:
    bit_count = len(hex_data) * 4
    return bin(int(hex_data, 16))[2:].zfill(bit_count)


def _bin2dec(bin_data: str) -> int:
    return int(bin_data, 2)


def _split_literal(bin_data: str, i: int) -> tuple[int, int]:
    log("literal")
    val = ""
    while True:
        proceed = bin_data[i]
        val += bin_data[i+1:i+5]
        i += 5
        if proceed == '0':
            break
    log(f"value: {val} ({_bin2dec(val)})")
    return val, i


def _split_operator(bin_data: str, i) -> tuple[list, int]:
    log("operator")
    length_type_id = bin_data[i]
    i += 1
    sub_packets = []
    if length_type_id == '0':
        size = _bin2dec(bin_data[i:i+15])
        log(f"size: {size}")
        i += 15
        cnt = 0
        while True:
            sub_packet, ii = _split(bin_data, i)
            cnt += ii - i
            log(f"used: {cnt}")
            i = ii
            if sub_packet is not None:
                sub_packets.append(sub_packet)
            if cnt == size:
                break
    elif length_type_id == '1':
        cnt = _bin2dec(bin_data[i:i+11])
        log(f"count: {cnt}")
        i += 11
        for j in range(cnt):
            sub_packet, i = _split(bin_data, i)
            sub_packets.append(sub_packet)
    else:
        raise ValueError("unexpected length_type_id")
    return (sub_packets, i)


def _split(bin_data: str, i: int) -> tuple[int, int, int, list]:
    tail = len(bin_data) - i
    if tail < 8:
        return (None, i + tail)
    version = _bin2dec(bin_data[i:i+3])
    log(f"version: {version}")
    type_id = _bin2dec(bin_data[i+3:i+6])
    log(f"type_id: {type_id}")
    i = i+6
    if type_id == 4:
        val, i = _split_literal(bin_data, i)
        return ((version, type_id, _bin2dec(val), []), i)
    else:
        operator_packets, i = _split_operator(bin_data, i)
        return ((version, type_id, None, operator_packets), i)


def _log_packet(packet: tuple[int, int, int, list]) -> None:
    log(packet)
    version, type_id, val, sub_packets = packet
    if type_id == 4:
        log("/".join([str(version), str(type_id), str(val)]))
    else:
        [_log_packet(p) for p in sub_packets]


def _sum_versions(packet: tuple[int, int, int, list]) -> int:
    version, _, _, sub_packets = packet
    return int(version) + sum(_sum_versions(p) for p in sub_packets)


def part_1(inputs: tuple[str]) -> int:
    assert len(inputs) == 1
    log(inputs[0])
    bin_data = _hex2bin(inputs[0])
    log(bin_data)
    packet, _ = _split(bin_data, 0)
    _log_packet(packet)
    log("")
    return _sum_versions(packet)


def _calc_value(packet: tuple[int, int, int, list]) -> int:
    _, type_id, _, sub_packets = packet
    values = [p[2] if p[2] else _calc_value(p) for p in sub_packets]
    if type_id == 0:
        return sum(values)
    elif type_id == 1:
        return prod(values)
    elif type_id == 2:
        return min(values)
    elif type_id == 3:
        return max(values)
    elif type_id == 5:
        assert len(values) == 2
        return values[0] > values[1]
    elif type_id == 6:
        assert len(values) == 2
        return values[0] < values[1]
    elif type_id == 7:
        assert len(values) == 2
        return values[0] == values[1]
    else:
        raise ValueError("unexpected type_id")


def part_2(inputs: tuple[str]) -> int:
    assert len(inputs) == 1
    bin_data = _hex2bin(inputs[0])
    packet, _ = _split(bin_data, 0)
    return _calc_value(packet)


TEST1 = "D2FE28".splitlines()
TEST2 = "38006F45291200".splitlines()
TEST3 = "EE00D40C823060".splitlines()
TEST4 = "C200B40A82".splitlines()
TEST5 = "04005AC33890".splitlines()
TEST6 = "880086C3E88112".splitlines()
TEST7 = "CE00C43D881120".splitlines()
TEST8 = "D8005AC2A8F0".splitlines()
TEST9 = "F600BC2D8F".splitlines()
TEST10 = "9C005AC2F8F0".splitlines()
TEST11 = "9C0141080250320F1802104A08".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2021, 16)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST1) == 6
    assert part_1(TEST2) == 9
    assert part_1(TEST3) == 14
    assert part_2(TEST4) == 3
    assert part_2(TEST5) == 54
    assert part_2(TEST6) == 7
    assert part_2(TEST7) == 9
    assert part_2(TEST8) == 1
    assert part_2(TEST9) == 0
    assert part_2(TEST10) == 0
    assert part_2(TEST11) == 1

    inputs = my_aocd.get_input(puzzle.year, puzzle.day, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == '__main__':
    main()

File: src/main/java/AoC2021_16.java
import static com.github.pareronia.aoc.CharArrayUtils.subarray;
import static java.util.stream.Collectors.toList;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.stream.LongStream;

import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.solution.Logger;
import com.github.pareronia.aoc.solution.LoggerEnabled;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2021_16 extends AoCBase {
    
    private static final class BITS {

        record Packet(
            Integer version,
            Integer typeId,
            Long value,
            List<Packet> subPackets
        ) {
            public static PacketBuilder builder() {
                return new PacketBuilder();
            }
            
            public static final class PacketBuilder {
                private Integer version;
                private Integer typeId;
                private Long value;
                private List<Packet> subPackets;
                
                public PacketBuilder version(final Integer version) {
                    this.version = version;
                    return this;
                }

                public PacketBuilder typeId(final Integer typeId) {
                    this.typeId = typeId;
                    return this;
                }

                public PacketBuilder subPackets(final List<Packet> subPackets) {
                    this.subPackets = subPackets;
                    return this;
                }

                public Packet build() {
                    return new Packet(version, typeId, value, subPackets);
                }
            }
        }
        
        public interface BITSEventHandler {
            default void startLiteralPacket(final long value) {
                // nop
            }
            
            default void startOperatorPacket(final int type) {
                // nop
            }
            
            default void version(final int version) {
                // nop
            }
            
            default void endLiteralPacket() {
                // nop
            }
            
            default void endOperatorPacket() {
                // nop
            }
        }
        
        public static class LoggingBITSHandler implements BITSEventHandler, LoggerEnabled {
            private final Logger logger;

            public LoggingBITSHandler(final Logger logger) {
                this.logger = logger;
            }

            @Override
            public Logger getLogger() {
                return this.logger;
            }

            @Override
            public void startLiteralPacket(final long value) {
                log(String.format("startLiteralPacket[value: %d]", value));
            }
            
            @Override
            public void startOperatorPacket(final int type) {
                log(String.format("startOperatorPacket[type: %d]", type));
            }
            
            @Override
            public void version(final int version) {
                log(String.format("version: %d", version));
            }

            @Override
            public void endLiteralPacket() {
                log("endLiteralPacket");
            }

            @Override
            public void endOperatorPacket() {
                log("endOperatorPacket");
            }
        }
        
        public static class DOMBITSHandler extends LoggingBITSHandler {
            private final Deque<Packet.PacketBuilder> builderStack = new ArrayDeque<>();
            private final Deque<Packet> packetStack = new ArrayDeque<>();
            
            public DOMBITSHandler(final Logger logger) {
                super(logger);
            }

            @Override
            public void startLiteralPacket(final long value) {
                super.startLiteralPacket(value);
                final Packet.PacketBuilder builder = Packet.builder();
                builder.typeId(4);
                builder.value = value;
                this.builderStack.push(builder);
            }

            @Override
            public void startOperatorPacket(final int type) {
                super.startOperatorPacket(type);
                final Packet.PacketBuilder builder = Packet.builder();
                builder.typeId(type);
                this.builderStack.push(builder);
            }

            @Override
            public void version(final int version) {
                super.version(version);
                this.builderStack.peek().version(version);
            }
            
            private void endPacket() {
                final Packet packet = builderStack.pop().build();
                if (builderStack.peek() != null) {
                    if (builderStack.peek().subPackets == null) {
                        builderStack.peek().subPackets(new ArrayList<>(List.of(packet)));
                    } else {
                        builderStack.peek().subPackets.add(packet);
                    }
                } else {
                    packetStack.push(packet);
                }
            }

            @Override
            public void endLiteralPacket() {
                super.endLiteralPacket();
                endPacket();
            }

            @Override
            public void endOperatorPacket() {
                super.endOperatorPacket();
                endPacket();
            }
            
            public Packet getPacket() {
                assert this.builderStack.size() == 0;
                assert this.packetStack.size() == 1;
                return this.packetStack.peek();
            }
        }
        
        public static final class Parser {
            private final BITSEventHandler handler;
            
            private Parser(final BITSEventHandler handler) {
                this.handler = handler;
            }

            public static Parser createParser(final BITSEventHandler handler) {
                return new Parser(handler);
            }
        
            public void parseHex(final String hexString) {
                parseBin(StringOps.hexToBin(hexString));
            }
            
            public void parseBin(final String binString) {
                parse(binString.toCharArray(), 0);
            }
            
            private int parse(final char[] binData, final int pos) {
                int i = pos;
                final int tail = binData.length - i;
                if (tail < 8) {
                    return i + tail;
                }
                final int version = binToDec(binData, i, i + 3);
                i += 3;
                final int typeId = binToDec(binData, i, i + 3);
                i += 3;
                if (typeId == 4) {
                    i = parseLiteral(binData, i);
                    handler.version(version);
                    handler.endLiteralPacket();
                } else {
                    handler.startOperatorPacket(typeId);
                    handler.version(version);
                    i = parseOperator(binData, i);
                    handler.endOperatorPacket();
                }
                return i;
            }
            
            private int parseLiteral(final char[] binData, final int pos) {
                int i = pos;
                final StringBuilder value = new StringBuilder();
                while (true) {
                    final char proceed = binData[i];
                    value.append(subarray(binData, i + 1, i + 5));
                    i += 5;
                    if (proceed == '0') {
                        break;
                    }
                }
                handler.startLiteralPacket(Long.parseLong(value.toString(), 2));
                return i;
            }

            private int parseOperator(final char[] binData, final int pos) {
                int i = pos;
                final int lengthTypeId = binData[i];
                i++;
                if (lengthTypeId == '0') {
                    final int size = binToDec(binData, i,  i + 15);
                    i += 15;
                    int cnt = 0;
                    while (true) {
                        final int ii = parse(binData, i);
                        cnt += ii - i;
                        i = ii;
                        if (cnt == size) {
                            break;
                        }
                    }
                } else if (lengthTypeId == '1') {
                    final int cnt = binToDec(binData, i, i + 11);
                    i += 11;
                    for (int j = 0; j < cnt; j++) {
                        i = parse(binData, i);
                    }
                } else {
                    throw new IllegalStateException("unexpected length_type_id");
                }
                return i;
            }

            private int binToDec(final char[] binData,
                            final int startInc, final int endExc) {
                final String s = String.valueOf(
                        subarray(binData, startInc, endExc));
                return Integer.parseInt(s, 2);
            }
        }
    }
    
    private final String hexData;
    
    private AoC2021_16(final List<String> input, final boolean debug) {
        super(debug);
        assert input.size() == 1;
        this.hexData = input.get(0);
    }
    
    public static final AoC2021_16 create(final List<String> input) {
        return new AoC2021_16(input, false);
    }

    public static final AoC2021_16 createDebug(final List<String> input) {
        return new AoC2021_16(input, true);
    }
    
    private static final class VersionSumBITSHandler extends AoC2021_16.BITS.LoggingBITSHandler {
        private int versionSum = 0;

        public VersionSumBITSHandler(final Logger logger) {
            super(logger);
        }

        public int getVersionSum() {
            return versionSum;
        }

        @Override
        public void version(final int version) {
            super.version(version);
            versionSum += version;
        }
    }
    
    private static final class ValueBITSCalcHandler extends AoC2021_16.BITS.DOMBITSHandler {

        public ValueBITSCalcHandler(final Logger logger) {
            super(logger);
        }
        
        private long calcValue(final AoC2021_16.BITS.Packet packet) {
            final List<Long> values = packet.subPackets().stream()
                .map(p -> {
                    if (p.value() != null) {
                        return p.value();
                    }
                    return calcValue(p);
                })
                .collect(toList());
            final LongStream longs = values.stream().mapToLong(Long::longValue);
            if (packet.typeId() == 0) {
                return longs.sum();
            } else if (packet.typeId() == 1) {
                return longs.reduce(1L, (a, b) -> a * b);
            } else if (packet.typeId() == 2) {
                return longs.min().orElseThrow();
            } else if (packet.typeId() == 3) {
                return longs.max().orElseThrow();
            } else if (packet.typeId() == 5) {
                assert values.size() == 2;
                return values.get(0) > values.get(1) ? 1L : 0L;
            } else if (packet.typeId() == 6) {
                assert values.size() == 2;
                return values.get(0) < values.get(1) ? 1L : 0L;
            } else {
                assert values.size() == 2;
                return values.get(0).equals(values.get(1)) ? 1L : 0L;
            }
        }
        
        public long getValue() {
            return calcValue(this.getPacket());
        }
    }
    
    @Override
    public Integer solvePart1() {
        final VersionSumBITSHandler handler = new VersionSumBITSHandler(this.logger);
        BITS.Parser.createParser(handler).parseHex(this.hexData);
        return handler.getVersionSum();
    }
    
    @Override
    public Long solvePart2() {
        final ValueBITSCalcHandler handler = new ValueBITSCalcHandler(this.logger);
        BITS.Parser.createParser(handler).parseHex(this.hexData);
        log(handler.getPacket());
        return handler.getValue();
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2021_16.create(List.of("D2FE28")).solvePart1() == 6;
        assert AoC2021_16.create(List.of("38006F45291200")).solvePart1() == 9;
        assert AoC2021_16.create(List.of("EE00D40C823060")).solvePart1() == 14;
        assert AoC2021_16.create(List.of("C200B40A82")).solvePart2() == 3;
        assert AoC2021_16.create(List.of("04005AC33890")).solvePart2() == 54;
        assert AoC2021_16.create(List.of("880086C3E88112")).solvePart2() == 7;
        assert AoC2021_16.create(List.of("CE00C43D881120")).solvePart2() == 9;
        assert AoC2021_16.create(List.of("D8005AC2A8F0")).solvePart2() == 1;
        assert AoC2021_16.create(List.of("F600BC2D8F")).solvePart2() == 0;
        assert AoC2021_16.create(List.of("9C005AC2F8F0")).solvePart2() == 0;
        assert AoC2021_16.create(List.of("9C0141080250320F1802104A08")).solvePart2() == 1;

        final Puzzle puzzle = Aocd.puzzle(2021, 16);
        puzzle.check(
            () -> lap("Part 1", () -> AoC2021_16.create(puzzle.getInputData()).solvePart1()),
            () -> lap("Part 2", () -> AoC2021_16.create(puzzle.getInputData()).solvePart2())
        );
    }
}